src/centroiders.cpp

```diff
@@ -1,95 +1,199 @@
+            yMax = i / imageWidth;
+        } else if (i / imageWidth < yMin) {
+            yMin = i / imageWidth;
+        }
+        magSum += image[i];
+        xCoordMagSum += ((i % imageWidth) + 1) * image[i];
+        yCoordMagSum += ((i / imageWidth) + 1) * image[i];
+        if((i + 1) % imageWidth != 0) {
+            cogHelper(yCoordMagSum, xCoordMagSum, magSum, xMin, xMax, yMin, yMax, cutoff, checkedIndeces, i + 1, image, imageWidth, imageHeight);
+        }
+        if ((i - 1) % imageWidth != (imageWidth - 1)) {
+            cogHelper(yCoordMagSum, xCoordMagSum, magSum, xMin, xMax, yMin, yMax, cutoff, checkedIndeces, i - 1, image, imageWidth, imageHeight);
+        }
+        cogHelper(yCoordMagSum, xCoordMagSum, magSum, xMin, xMax, yMin, yMax, cutoff, checkedIndeces, i + imageWidth, image, imageWidth, imageHeight);
+        cogHelper(yCoordMagSum, xCoordMagSum, magSum, xMin, xMax, yMin, yMax, cutoff, checkedIndeces, i - imageWidth, image, imageWidth, imageHeight);
+    }
+}
+
+std::vector<Star> CenterOfGravityAlgorithm::Go(unsigned char *image, int imageWidth, int imageHeight) const {
+    float yCoordMagSum = 0; 
+    float xCoordMagSum = 0;
+    int magSum = 0;
+    int xMin;
+    int xMax;
+    int yMin;
+    int yMax;
+    int cutoff;
+    std::unordered_set<int> checkedIndeces;
+    
+    std::vector<Star> result;
+    //loop through entire array, find sum of magnitudes
+    int totalMag = 0;
+    for (int i = 0; i < imageHeight * imageWidth; i++) {
+        totalMag += image[i];
+    }
+    // cutoff might need a new equation
+    cutoff = (((totalMag/(imageHeight * imageWidth)) + 1) * 15) / 10;
+    for (int i = 0; i < imageHeight * imageWidth; i++) {
+        //check if pixel is part of a "star" and has not been iterated over
+        if (image[i] >= cutoff && checkedIndeces.count(i) == 0) {
+            checkedIndeces.insert(i);
+            //iterate over pixels that are part of the star
+
+            int xDiameter = 0; //radius of current star
+            int yDiameter = 0;
+            yCoordMagSum = 0; //y coordinate of current star
+            xCoordMagSum = 0; //x coordinate of current star
+            magSum = 0; //sum of magnitudes of current star
+
+            //computes indices to skip after done w current star
+            int j = i;
+            while(j < imageWidth * imageHeight && (j + 1) % imageWidth != 0 && image[j] >= cutoff) {
+                j++;
+            }
+
+            magSum += image[i];
+            xMax = i % imageWidth;
+            xMin = i % imageWidth;
+            yMax = i / imageWidth;
+            yMin = i / imageWidth;
+            xCoordMagSum += ((i % imageWidth) + 1) * image[i];
+            yCoordMagSum += ((i / imageWidth) + 1) * image[i];
+            if((i + 1) % imageWidth != 0) {
+                cogHelper(yCoordMagSum, xCoordMagSum, magSum, xMin, xMax, yMin, yMax, cutoff, checkedIndeces, i + 1, image, imageWidth, imageHeight);
+            }

```

Comment:
Can we simplify this into a single call of `cogHelper(..., i, ...)` rather than reimplementing the recursion logic from `cogHelper`?

Reply:
yes we can

Reply:
Done - Main now calls cogHelper() only once, everything still works fine